### ISSUE TITLE
Bazelisk: error during the fetch of repository 'google_bazel_common'

### ISSUE THREAD

**Issue by Getfree** [google/closure-compiler#3717]
I'm trying to build Closure-Compiler v20201006 using Bazelisk under Windows 7 x64.

I get the following error:
```
A:\Desktop\closure-compiler-20201006> ..\bazelisk build //:compiler_unshaded_deploy.jar
Loading:
Loading: 0 packages loaded
INFO: Call stack for the definition of repository 'google_bazel_common' which is a git_repository (rule definition at C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git.bzl:195:18):
 - <builtin>
 - A:/desktop/closure-compiler-20201006/WORKSPACE.bazel:5:1
ERROR: An error occurred during the fetch of repository 'google_bazel_common':
   Traceback (most recent call last):
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git.bzl", line 177
                _clone_or_update(ctx)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git.bzl", line 36, in _clone_or_update
                git_repo(ctx, directory)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 91, in git_repo
                _update(ctx, git_repo)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 101, in _update
                init(ctx, git_repo)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 115, in init
                _error(ctx.name, cl, st.stderr)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 181, in _error
                fail(<1 more arguments>)
error running 'git init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common' while working with @google_bazel_common:
java.io.IOException: ERROR: src/main/native/windows/process.cc(199): CreateProcessW("git" init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common): The system cannot find the file specified.
 (error: 2)
ERROR: no such package '@google_bazel_common//': Traceback (most recent call last):
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git.bzl", line 177
                _clone_or_update(ctx)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git.bzl", line 36, in _clone_or_update
                git_repo(ctx, directory)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 91, in git_repo
                _update(ctx, git_repo)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 101, in _update
                init(ctx, git_repo)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 115, in init
                _error(ctx.name, cl, st.stderr)
        File "C:/users/administrator/_bazel_<user>/zflckzzj/external/bazel_tools/tools/build_defs/repo/git_worker.bzl", line 181, in _error
                fail(<1 more arguments>)
error running 'git init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common' while working with @google_bazel_common:
java.io.IOException: ERROR: src/main/native/windows/process.cc(199): CreateProcessW("git" init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common): The system cannot find the file specified.
 (error: 2)
INFO: Elapsed time: 0.542s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
FAILED: Build did NOT complete successfully (0 packages loaded)
```

I'm not certain who is to blame here.
Is it Closure Compiler's fault?
Is it Bazelisk's fault?
Is it Bazel's fault?

The output of `bazelisk version` is:
```
Bazelisk version: v1.7.4
Build label: 2.2.0
Build target: bazel-out/x64_windows-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Tue Mar 3 09:28:29 2020 (1583227709)
Build timestamp: 1583227709
Build timestamp as int: 1583227709
```

**Comment by mprobst** [comment]
```
error running 'git init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common' while working with @google_bazel_common:
java.io.IOException: ERROR: src/main/native/windows/process.cc(199): CreateProcessW("git" init C:/users/administrator/_bazel_<user>/zflckzzj/external/google_bazel_common): The system cannot find the file specified.
```

This sounds like a setup issue. Do you have a working `git` binary installed?

**Comment by Getfree** [comment]
I didn't know I needed that. Git is not mentioned as prerequisite in the build instructions.
It turns out that I also need MSYS2, which is also not mentioned as prerequisite.

I installed those tools and now I get the following error:
```
A:\Desktop\closure-compiler-20201102> bazelisk build //:compiler_unshaded_deploy.jar
WARNING: Running Bazel server needs to be killed, because the startup options are different.
Starting local Bazel server and connecting to it...
Loading:
Loading: 0 packages loaded
Loading: 0 packages loaded
Loading: 0 packages loaded
    currently loading:
Analyzing: target //:compiler_unshaded_deploy.jar (1 packages loaded)
Analyzing: target //:compiler_unshaded_deploy.jar (1 packages loaded, 0 targets configured)
Analyzing: target //:compiler_unshaded_deploy.jar (9 packages loaded, 11 targets configured)
Analyzing: target //:compiler_unshaded_deploy.jar (11 packages loaded, 18 targets configured)
Analyzing: target //:compiler_unshaded_deploy.jar (11 packages loaded, 18 targets configured)
Analyzing: target //:compiler_unshaded_deploy.jar (16 packages loaded, 249 targets configured)
Analyzing: target //:compiler_unshaded_deploy.jar (30 packages loaded, 375 targets configured)
INFO: Analyzed target //:compiler_unshaded_deploy.jar (51 packages loaded, 2196 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base C:/users/administrator/_bazel_<user>/cy7rdti4/sandbox
[0 / 16] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[4 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/code_generator.cc [for host]; 4s local ... (2 actions, 1 running)
[6 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_doc_comment.cc [for host]; 3s local ... (2 actions, 1 running)
[7 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_enum_field.cc [for host]; 5s local ... (2 actions running)
[9 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_message.cc [for host]; 4s local ... (2 actions running)
[10 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_message.cc [for host]; 11s local ... (2 actions running)
[10 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_message.cc [for host]; 20s local ... (2 actions running)
[12 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_extension_lite.cc [for host]; 2s local ... (2 actions running)
[15 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_file.cc [for host]; 4s local ... (2 actions running)
[18 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/csharp/csharp_wrapper_field.cc [for host]; 4s local
[22 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_map_field_lite.cc [for host]; 6s local
[26 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_message_field.cc [for host]; 6s local ... (2 actions running)
[32 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_string_field.cc [for host]; 1s local ... (2 actions running)
[37 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_generator.cc [for host]; 6s local ... (2 actions running)
[43 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/php/php_generator.cc [for host]; 8s local ... (2 actions running)
[50 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/csharp/csharp_message.cc [for host]; 8s local ... (2 actions running)
[58 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_file.cc [for host]; 8s local ... (2 actions running)
[64 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_helpers.cc [for host]; 9s local ... (2 actions running)
[71 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/cpp/cpp_service.cc [for host]; 3s local ... (2 actions running)
[84 / 143] Compiling external/com_google_protobuf/src/google/protobuf/compiler/java/java_name_resolver.cc [for host]; 1s local ... (2 actions running)
[116 / 196] Compiling external/com_google_protobuf/src/google/protobuf/util/internal/type_info.cc [for host]; 5s local ... (2 actions running)
[130 / 196] Compiling external/com_google_protobuf/src/google/protobuf/struct.pb.cc [for host]; 1s local ... (2 actions, 1 running)
[157 / 196] Compiling external/com_google_protobuf/src/google/protobuf/util/time_util.cc [for host]; 5s local ... (2 actions, 1 running)
ERROR: A:/desktop/closure-compiler-20201102/BUILD.bazel:151:1: Executing genrule //:externs_zip failed (Exit 127)
Target //:compiler_unshaded_deploy.jar failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 702.918s, Critical Path: 51.93s
INFO: 170 processes: 170 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```
It says `use  --verbose_failures` near the end. So I did that and I got this:
```
A:\Desktop\closure-compiler-20201102> bazelisk build //:compiler_unshaded_deploy.jar --verbose_failures
Loading:
Loading: 0 packages loaded
Analyzing: target //:compiler_unshaded_deploy.jar (0 packages loaded, 0 targets configured)
INFO: Analyzed target //:compiler_unshaded_deploy.jar (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
[0 / 10] [Prepa] BazelWorkspaceStatusAction stable-status.txt
[1 / 25] Action external/bazel_tools/tools/jdk/platformclasspath.jar; 1s local ... (2 actions running)
ERROR: A:/desktop/closure-compiler-20201102/BUILD.bazel:151:1: Executing genrule //:externs_zip failed (Exit 127): bash.exe failed: error executing command
  cd C:/users/administrator/_bazel_<user>/cy7rdti4/execroot/com_google_javascript_jscomp
  SET PATH=C:\tools\msys64\usr\bin;C:\tools\msys64\bin;C:\Windows;C:\Windows\System32;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Users\Administrator\AppData\Local\bazelisk\downloads\bazelbuild\bazel-2.2.0-windows-x86_64\bin;C:\ProgramData\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\WindowsPowerShell\v1.0\;A:\App files\Node.js v8.12.0\;A:\App files\Git\cmd
    SET RUNFILES_MANIFEST_ONLY=1
  C:/tools/msys64/usr/bin/bash.exe -c source external/bazel_tools/tools/genrule/genrule-setup.sh; cd externs && find . -name '*.js' | zip ../bazel-out/x64_windows-fastbuild/bin/externs.zip -@ &> /dev/null
Execution platform: @local_config_platform//:host
Target //:compiler_unshaded_deploy.jar failed to build
INFO: Elapsed time: 5.224s, Critical Path: 2.54s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

It says that `bash.exe failed: error executing command`, but there are many commands being shown there. I'm clueless as to what specific command is failing.

**Comment by mprobst** [comment]
I think the `zip` command is what is failing. Do you have a `zip` binary on
your system?

Taking a step back, what problem are you actually trying to solve here?

**Comment by Getfree** [comment]
I found a linux-compatible zip.exe here: http://stahlworks.com/dev/index.php?tool=zipunzip
I added it to the PATH and now the build process completed successfully.

Where I work they use a custom version of Closure Compiler that was built from version 20190121.
There are several bugs in that version that have been fixed in these two years, so I needed to rebuild that custom Closure Compiler using the latest version as a base.

Bottom line, the missing prerequisites are:
- Git
- MSYS2
- A zip command-line executable (compatible with the one in linux)

It will help others if you specify them in the build instructions.
Thanks for your help.

**Comment by mprobst** [comment]
Good point, and thanks for researching the tooling requirements.

**Comment by DerGernTod** [comment]
@Getfree i know it's been a while, but i'm having problems still with the same command. i did everything you mentioned (git, msys2, that zip tool, adding it to PATH etc) but i still get the same error
```
BUILD.bazel:186:8: Executing genrule //:externs_zip failed: (Exit 127): bash.exe failed: error executing command
```
do you remember doing anything else?

**Comment by Getfree** [comment]
When I did the entire process back then, I wrote the instructions step by step in case I needed to do it again.
Here they are:

1. Download and uncompress the newest closure-compiler-20201102 file from https://github.com/google/closure-compiler/releases
	- Uncompress it to a folder path without spaces (bazelisk complains about it)

2. Download and install Java SDK (install just the SDK, not the source code or the JRE or anything else)

3. Set the JAVA_HOME environment variable to C:\Program Files\Java\jdk1.8.0_201

4. Download the latest bazelisk binary from https://github.com/bazelbuild/bazelisk/releases
	- rename it to "bazelisk.exe" and move it to the closure-compiler-20201102 folder
	
5. Download a portable Git binary from https://git-scm.com/download/win
	- Uncompress it to "A:\App files\Git"
	- Add "A:\App files\Git\cmd" to the PATH variable (Control Panel > User Accounts > Change my environment variables)

6. Download and install MSYS2 from https://www.msys2.org

7. Download zip.exe from http://stahlworks.com/dev/index.php?tool=zipunzip (direct link http://stahlworks.com/dev/zip.exe)
	- Add zip.exe's path to the PATH environment variable

8. Open a cmd window under closure-compiler-20201102 and run the command:
	bazelisk build //:compiler_unshaded_deploy.jar
	- The command takes about 7 minutes the first time. And about 1 minute the second time on.
	
9. Copy the file "closure-compiler-20201102\bazel-bin\compiler_unshaded_deploy.jar" to wherever it's needed.

**Comment by DerGernTod** [comment]
thanks a lot! did everything in your list but the externs_zip command still failed. found out why, it's really the `find` command. per default it takes the one from cmd which works completely different. i pointed it to the msys64 one within BUILD.bazel and now it works. i still have another issue when trying to build `:all` though (`//:compiler_unshaded_deploy.jar` works fine): 
```
bazel-out/x64_windows-fastbuild/bin/compiler_lib_no_runtime_libs.javadoc.run_shell_0.sh: line 2: external/local_jdk/bin/javadoc: Argument list too long
```

other than that, not wanting to go too much into smalltalk, but did you manage to get the project into eclipse or vscode to debug it?

**Comment by Getfree** [comment]
I remember I did open the project in vscode, but my debugging consisted of simple log lines to a text file.
i.e. I didn't use breakpoints, step by step or anything like that.
